### ISSUE TITLE
Facet search specification

### ISSUE THREAD

**Issue by ayco-at-naturalis** [Sp2000/colplus-backend#157]
The facet search facility is modeled (at least initially) after the GBIF species search page. See https://www.gbif.org/species/search

For each of the distinct values within a facet, a document count must be provided. For example the "dataset" facet has values like "Zoobank", "ITIS", "TAXREF", etc. Zoobank has X name usages, ITIS has Y name usages and TAXREF has Z name usages.

Each facet also functions as an extra query constraint besides the search phrase entered by the user. For example, the user may search for "lepidoptera" and also check the checkbox for the Zoobank dataset, so he will search for all lepidopera within the Zoobank dataset.

If the user selects another value within the **same** facet, this results in an OR query - e.g. search for lepidotera in EITHER the ZooBank dataset OR the ITIS dataset.

If the user selects a value in **another** facet, this results in an AND query:
```
q="lepidopera"
AND (dataset="ZooBank" OR dataset="ITIS")
AND (rank="species" OR rank="subspecies")

```
However, selecting values within a facet **also** impacts the document counts for the values in the other facets. If you check the Zoobank checkbox, the document counts in the rank facet must be recalculated (show me the number of subspecies, species, genera, etc. just for the Zoobank dataset). Conversely, if you check the "species" checkbox within the rank facet, the document counts for the dataset facet must be recalculated (just show me the number of species-level documents in Zoobank, ITIS, etc).

In other words, selecting or deselecting any value within a facet requires the recalculation of document counts for all other facets.

**Comment by mdoering** [comment]
Thanks @ayco-at-naturalis. Selecting a facet value causes a new search to be issued which returns new counts obviously. Just the facet which already includes a filter should do the counts as if the facet was unselected. I think solr does that by default, not sure how to do that in ES best.

Sidenote as for values: the dataset names are just in the UI, underlying facets use the datasetKey just like we have in CoL.

Also selecting multiple values for the same field/search parameter results in an OR, while filters across fields are combined with AND

**Comment by ayco-at-naturalis** [comment]
I was also wondering whether ES had such classic prefab facet functionality ... I seem to not have noticed it ... but I also wonder whether the description I gave (with the mutually updating facet counts) really is totally classic facet functionality (thinking of web shops and such)

**Comment by ayco-at-naturalis** [comment]
I suspect the "filters" aggregation serves our purpose prettu accurately https://www.elastic.co/guide/en/elasticsearch/reference/6.4/search-aggregations-bucket-filters-aggregation.html

**Comment by mdoering** [comment]
Not so sure, but I am really not a ES expert and used to Solr terminology.
Seems to me that you need to tell the bucket filters the values you want the filters to work with.

In our case we want to list the fields indexed as keywords that we like to list with all their values. Dunno.

**Comment by mdoering** [comment]
@ayco-at-naturalis is this complete and can be closed? 
Please provide here some working API links that show the facets workin on prod

**Comment by ayco-at-naturalis** [comment]
Here some working example URLs:

http://api-dev.col.plus/name/search/?facet=rank&status=accepted
Only see facets
http://api-dev.col.plus/name/search/?facet=rank&facet=status&status=misapplied&facet=issue&limit=0

Once higher taxon id search is integrated, this one works too (works with me locally already):
http://api-dev.col.plus/name/search/?facet=taxon_id&rank=order

**Comment by ayco-at-naturalis** [comment]
While coming with examples, I stumpled over a bug with integer facets (basically all ***_key params). That's now solved. We do need to re-index everything though.
